<?xml version="1.0"?>
<launch>
  ###########  launch config  ###########
  <arg name="real_machine" default="True" />
  <arg name="simulation" default="False" />
  <arg name="control_mode" default="0" />
  <arg name="type" default="quad" />
  <arg name="estimate_mode"  default= "2" />
  <arg name="headless" default="True" />

  <!-- sensors(fusion) special parameters -->
  <param name="/sensor_plugin/altitude/no_height_offset" value="true" />
  <param name="/sensor_plugin/mocap/cutoff_vel_freq" value="10" />

  ###########  UAV Config  ###########
  # DRAGON = 32 # 0x20, dragon type with full/over actuated property
  <param name="/uav_info/uav_model" value= "32" />

  ###########  Motor Config  ###########
  <rosparam file="$(find dragon)/config/MotorInfo.yaml" command="load" />

  ###########  Battery Config  ###########
  <rosparam file="$(find dragon)/config/$(arg type)/Battery.yaml" command="load" />

  ###########  Base Platform  ###########
  <node pkg="aerial_robot_base" type="aerial_robot_base_node" name="aerial_robot_base_node" output="screen" >

    ###########  Basic Param  ###########
    # EGOMOTION_ESTIMATE = 0
    # EXPERIMENT_ESTIMATE = 1. for unstable mocap, use this mode
    # GROUND_TRUTH = 2
    <param name="estimator/estimate_mode" value= "$(arg estimate_mode)" if="$(arg real_machine)"/>
    <param name="estimator/estimate_mode" value= "2" if="$(arg simulation)" />

    <param name="navigator/xy_control_mode"  value="$(arg control_mode)"/>

    <param name="simulation" value="$(arg simulation)" />
    <param name="param_verbose" value= "true" />
    <param name="main_rate" type="double" value="40"/>

    ###########  Sensor Fusion  ###########
    <rosparam file="$(find dragon)/config/SensorFusion.yaml" command="load" if="$(arg real_machine)"/>
    <rosparam file="$(find dragon)/config/SensorSimulation.yaml" command="load" if="$(arg simulation)"/>
    ###########  Gimbal Control  ###########
    <param name="control_plugin_name" value ="control_plugin/dragon_gimbal" />
    <rosparam file="$(find dragon)/config/$(arg type)/GimbalControlConfig.yaml" command="load" />
    <remap from="/gimbals_ctrl" to="/dragon/gimbals_ctrl"/>
    <remap from="/joints_ctrl" to="/dragon/joints_ctrl"/>
    <remap from="/joint_states" to="/dragon/joint_states"/>

    ###########  Teleop  ###########
    <rosparam file="$(find dragon)/config/TeleopNavigationConfig.yaml" command="load" />
  </node>

  ###########  Sensors  ###########
  <include file="$(find aerial_robot_base)/launch/sensors/onboard_sensors.launch"  if="$(arg real_machine)" >
    <arg name="imu_board" value="1" />  <!-- 1: d imu board, 0: kduino board -->
    <arg name="use_alt" value="0" />
    <arg name="use_mocap" value="0" />
    <arg name="use_joy" value="0" />
    <arg name="use_opti" value="0" />
    <arg name="imu_port" value="/dev/ttyUSB0" />
    <arg name="alt_port" value="/dev/leddar_one" />
  </include >

  ###########  Servo Bridge  ###########
  <node pkg="dragon" type="dynamixel_bridge_node" name="servo_bridge"  output="screen"  ns="dragon"  if="$(arg real_machine)" >
    <rosparam file="$(find dragon)/config/$(arg type)/Servo.yaml" command="load" />
  </node>


  ###########  Robot Model  ###########
  <include file="$(find dragon)/launch/robot_model.launch">
    <arg name="type" value="$(arg type)" />
    <arg name="headless" value="$(arg headless)" />
    <arg name="need_joint_state" value="false" if="$(arg real_machine)"/>
    <arg name="need_joint_state" value="false" if="$(arg simulation)"/>
    <arg name="description_mode" value="gazebo" if="$(arg simulation)"/>
  </include >

  ###########  Transform Control  ###########
  <node pkg="dragon" type="dragon_transform_control_node" name="dragon_transformcontrol"  output="screen" respawn="true">
    <rosparam file="$(find dragon)/config/$(arg type)/AttitudeControl.yaml" command="load" />
    <remap from="joint_states" to="/dragon/joint_states"/>
    <param name="kinematic_verbose" value="false" />
    <param name="control_verbose" value="false" />
    <param name="debug_verbose" value="false" />
    <param name="real_machine" value="$(arg real_machine)" />
  </node>

  ########## Simulation in Gazebo #########
  <rosparam file="$(find dragon)/config/$(arg type)/SimulationControl.yaml" command="load"/>
  <include file="$(find aerial_robot_simulation)/launch/simulation.launch" if = "$(arg simulation)" >
    <arg name="model" default="dragon" />
    <arg name="gui" default="false" if="$(arg headless)" />
    <arg name="headless" default="$(arg headless)" />
  </include>
</launch>



